### ISSUE TITLE
Fixing code coverage

### ISSUE THREAD

**Issue by alex-konovalov** [DistanceDevelopment/Distance#90]
Now, with the basic CI in place after #87, we may try to get code coverage to work again. I see that it used to work in the past, but stopped updating https://app.codecov.io/gh/DistanceDevelopment/Distance about 10 months ago.

@dill I've tried it before from https://github.com/r-lib/actions/blob/master/examples/test-coverage.yaml and it did not work from the first run. I suggest now we should proceed in a pull request instead of making commits to the master branch. The idea is that you can edit pull request as many times as you wish, until it will be ready for merging ("merge with rebase" preferably, for a cleaner commits history). This way, we will avoid making commits to the master only to see that they fail tests and need to be fixed again. To the contrary, in a branch from which a PR is made, it's ok to have multiple so called "fixup" commits, because you can then rewrite the history before merging it (e.g. with "squash and merge" in GitHub web-interface, or even better using Git in the command line locally).

**Comment by dill** [comment]
Thanks @alex-konovalov, I didn't know that the PR would allow this, that's a great tip.

I'll close this for now but will keep this in mind as I try to get the coverage working.

**Comment by alex-konovalov** [comment]
As you'd like :) Although you could submit a PR or a commit and [use magic word to close an issue automatically](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), to make a better trail. 

In https://github.com/alex-konovalov/tempConverter/pull/3 @LHMarshall set up code coverage for a toy project we have created to try out things - you may have a look.

**Comment by alex-konovalov** [comment]
@dill Tried in my fork. The same setup that works for `tempConverter` fails. Some googling suggested that a problem may be in test files. Removed all test files except one, and the output is now more verbose: see https://github.com/alex-konovalov/Distance/runs/2788627186?check_suite_focus=true
```
rror: non-conformable arguments
Backtrace:
    █
 1. └─Distance::ds(...) test_bootdht.R:8:2
 2.   └─mrds::check.mono(model, n.pts = 20)
 3.     └─base::apply(udat, 1, chpply, x = x, strict = strict, ddfobj = ddfobj)
 4.       └─mrds:::FUN(newX[, i], ...)
 5.         ├─base::as.vector(detfct(x, ddfobj, width = right.trunc, standardize = TRUE))
 6.         └─mrds::detfct(x, ddfobj, width = right.trunc, standardize = TRUE)
 7.           └─mrds:::scalevalue(ddfobj$scale$parameters, scale.dm)
 8.             └─base::as.vector(as.matrix(z) %*% key.scale)
```
Does this give any clue?

**Comment by dill** [comment]
Thanks Alex, this is helpful. This is a bug in `mrds` that I have fixed already, so I think the issue is that [when installing the `Distance` dependencies](https://github.com/alex-konovalov/Distance/actions/runs/923368983/workflow#L46) we install the CRAN version of `mrds`, not the dev version. I'll see if I can get somewhere with this...

**Comment by alex-konovalov** [comment]
@dill wonderful! I've added a line
```
          remotes::install_github("DistanceDevelopment/mrds")
```
and now it passes https://github.com/alex-konovalov/Distance/runs/2792052453?check_suite_focus=true (with one file). Let me rebase to see if the rest of tests will also pass. Are there any other packages like mrds, where you may need GitHub version?

**Comment by dill** [comment]
Most likely I'll need to do the same with dsm, so that mrds and Distance 
github versions are used but that's not on my schedule for a few weeks.

Not sure if this is also the case for some of @LHMarshall's packages.

**Comment by alex-konovalov** [comment]
Sorry, I meant "any other packages that Distance needs to install from GitHub to make its test pass". I've put back all other test files, and there are again failures.

**Comment by alex-konovalov** [comment]
@dill I have identified one file which causes it - removing tests/testthat/test_ds.R  in my fork as an experiment, and then coverage passes: https://github.com/alex-konovalov/Distance/runs/2793329715?check_suite_focus=true

**Comment by dill** [comment]
I should have reported back here too, that's where I've got to as well. 
I think this is down to some numerical differences. I'm trying to 
minimize the number of tests that reply on numerical agreement between 
platforms which I think is the crux of this issue. I'm not going to get 
back to this for a few weeks now due to conference/leave but I'm glad 
we've found the problem at least :)

**Comment by alex-konovalov** [comment]
@dill nice! At least the R-CMD tests are now passing!

We need to find a way to run tests so that we can have more informative test logs. Are the same tests run as a part of R_CMD tests, but without coverage reports? Or rather not? In the latter case I see the ideal workflow as
1) R-CMD
2) "testthat" tests
3) coverage reports

If (3) fails, and (2) fails, then look at (2) and see where it failed (in case it's the nature of coverage to be less informative due to wrapping the process into some environment to trace lines of code used).

P.S. Maybe reopen this issue - really not comfortable to track discussions in closed ones :)

**Comment by dill** [comment]
Re-opened, agree :)

This is a bit more complicated. By default tests are run when running `R CMD check` but I exclude them in package building (using `.Rbuildignore`) so that they are not run on CRAN, which was previously causing a headache (so I think I'd set something up to run the tests on Travis-CI separately).

I think this is a reasonable workflow. To get a bit more complicated, I think we can run an action on successful completion of a previous action, so if that kind of cascading structure can work, we would be able to have `R CMD check` simply check if the package can be installed/built (via `devtools::check(run_dont_test=TRUE)`, say), then if it can run tests and if the tests work only then run coverage.

**Comment by alex-konovalov** [comment]
Ah, that explains why this differs from https://github.com/alex-konovalov/tempConverter/ :) Thanks for explaining that. Should be possible to find a way to have two sets of tests, one for CRAN, one for GitHub actions. Cascading may be a good idea, but in this case if one works on a pull requests and test fails because of some trivial things in R CMD, they will not know if the rest of the tests pass or how good the coverage of the modified code is, so after all it's not expensive to run the whole workflow, and we should not worry too much...

**Comment by alex-konovalov** [comment]
P.S. Perhaps getting tests run on GitHub actions is now more important than getting coverage. My other suggestion is to try to use more pull requests to get them tested *prior* to merge; otherwise, you commit to master, test fails, in the meantime, while you're fixing it, users install package with `install_github("DistanceDevelopment/Distance")` and get broken version.

**Comment by dill** [comment]
Thanks again @alex-konovalov. Closing here as we've merged but I'll open another issue regarding `skip_on_cran()`
